### PR TITLE
Expose ArchetypeGenerator and ArchetypeManager

### LINKED ISSUE #921: API missing for creating archetype projects
Previously there was the public `IProjectConfigurationManager.createArchetypeProjects(...)` but now there is only the internal class `ArchetypeGenerator` (introduced with https://github.com/eclipse-m2e/m2e-core/commit/eb408339868ce4484e52b2bf9f5df59b5ea76229#diff-4742ec5a022dec0edc843e783c76a0a258224fcb273fb08ba3f107e0dc6887cb). How should consumers now trigger a project creation for archetypes?

Was the `ArchetypeGenerator` being put deliberately in package `org.eclipse.m2e.core.ui.internal.archetype`?

### PR DESCRIPTION
**Pull request by kwin**
This closes #921

### CODE REVIEW

**Comment by kwin** [comment]
@laeubi The compilation inside Eclipse works fine and `org.eclipse.m2e.core.ui` list `org.osgi.annotation.versioning_1.1.2.202109301733.jar` as plug-dependency. Still Tycho cannot compile and fails with

```
Error:  Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:2.7.4:compile (default-compile) on project org.eclipse.m2e.core.ui: Compilation failure: Compilation failure: 
Error:  /home/runner/work/m2e-core/m2e-core/org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/wizards/MavenDependenciesWizardPage.java: 
Error:  	/*******************************************************************************
Error:  	^
Error:  The type org.osgi.annotation.versioning.Version cannot be resolved. It is indirectly referenced from required .class files
Error: [ERROR] 1 problem (1 error)
Error:  -> [Help 1]
```

The target file though contains the bundle `org.osgi.annotation.versioning` so I am puzzled why Tycho does not pick up that dependency....

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import org.eclipse.m2e.core.project.IArchetype;
+import org.eclipse.m2e.core.project.MavenProjectInfo;
+
+
+/**
+ * Implementations of this OSGi service interface allow to create new projects based on Maven archetypes.
+ * 
+ * @since 2.0.1 (package version 1.0.0)
```
Comment (mickaelistria) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:32]: Will be a 2.1 for new APIs, not 2.0.1

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/*******************************************************************************
```
Comment (mickaelistria) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: We usually don't use package-info.json. Is it something really needed and useful here? If not, I suggest we remove it and maybe add it later, when there is a compelling reason to add them everywhere in m2e.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java [hunk 3]
```diff
@@ -64,19 +65,10 @@ public class ArchetypeGenerator {
...
   IMavenLauncher mavenLauncher;
 
-  /**
-   * Creates project structure using Archetype and then imports created project(s)
-   *
-   * @return a list of created projects.
-   * @since 1.8
+  /* (non-Javadoc)
+   * @see org.eclipse.m2e.core.ui.internal.archetype.ArchetypeGeneratorInterface#createArchetypeProjects(org.eclipse.core.runtime.IPath, org.eclipse.m2e.core.project.IArchetype, java.lang.String, java.lang.String, java.lang.String, java.lang.String, java.util.Properties, boolean, org.eclipse.core.runtime.IProgressMonitor)
    */
```
Comment (mickaelistria) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java:70]: This is not necessary and should be removed. @Override is enough here to reference documentation from parent type.

**Comment by laeubi** [comment]
> The target file though contains the bundle `org.osgi.annotation.versioning` so I am puzzled why Tycho does not pick up that dependency....

I think the problem is that Tycho 2.4 is used here.

**Review by mickaelistria — CHANGES_REQUESTED** [review]
Just moving this is IMO not enough. We need to audit what are the real operations necessary to be exposed for clients to efficiently use the Archetype API here. Many operations in the promoted APIs can probably remain implementation-internal and should then be so.

**Comment by kwin** [comment]
> Just moving this is IMO not enough. We need to audit what are the real operations necessary to be exposed for clients to efficiently use the Archetype API here.

The reasoning is in https://github.com/eclipse-m2e/m2e-core/issues/921. This PR obviously just contains the code changes.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java (hunk 2 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: This is `package-info.java` not `.json`. It is necessary to version on package level as recommended by OSGi. Otherwise you cannot do proper semantic versioning as outlined in https://docs.osgi.org/whitepaper/semantic-versioning/040-semantic-versions.html.

Comment (mickaelistria) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: > It is necessary to version on package level as recommended by OSGi. Otherwise you cannot do proper semantic versioning as outlined in https://docs.osgi.org/whitepaper/semantic-versioning/040-semantic-versions.html.

package version has to be declared in MANIFEST.MF, not with a package-info.java. There might be some tools to build a good MANIFEST with this file, but the MANIFEST-first approach used in m2e doesn't allow that and versions need to be set directly in the MANIFEST.MF Export-Package.

Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: Is the ConsumerType/ProviderType annotations evaluated by PDE for automatically generating the correct import-version ranges? Are there any plans for Tycho/PDE to evaluate the official annotations?
I will update the PR accordingly to export this package in version 1.0.0 manually in MANIFEST.MF.

**Comment by mickaelistria** [comment]
> The reasoning is in https://github.com/eclipse-m2e/m2e-core/issues/921.

The reasoning "it was this way before" is not viable. The API was removed in this major version because it wasn't appearing as useful to anyone involved regularly in the development of m2e. That is already a sign that the API state should be challenged and not cloned for the future of m2e. If we're to reopen an API, it can and should be a minimal one, just like if we were starting it from scratch.
What is the particular methods you're using in your code? We should just start by exposing as API and keep everything else internal until we have a good story to promote them.

**Comment by kwin** [comment]
https://github.com/eclipse-m2e/m2e-core/issues/921#issuecomment-1253955753
> The use case is triggering new project creation based on a specific artifact coordinates (https://github.com/apache/sling-ide-tooling/blob/93bcb69fe7cfad6b68d5caf838a14d3bdf887b92/eclipse/eclipse-m2e-ui/src/org/apache/sling/ide/eclipse/ui/wizards/np/AbstractNewMavenBasedSlingApplicationWizard.java#L106).

**Comment by mickaelistria** [comment]
So you would only need a way to run something like `archetypeGenerator.createProject(Archetype, ...)` ?

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java (hunk 2 above)
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: > package version has to be declared in MANIFEST.MF, not with a package-info.java. There might be some tools to build a good MANIFEST with this file, but the MANIFEST-first approach used in m2e doesn't allow that and versions need to be set directly in the MANIFEST.MF Export-Package.

As noted above this is not true anymore PDE already handles these annotations and updates the manifest automatically.
Beside that, it allows other tools to find out what is the purpose of an interface if it is used as a dependency so using these annotation can (and should be) done to move forward towards best OSGi practice.

Comment (mickaelistria) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java:1]: Good to know that modification can be automated, but still the MANIFEST.MF does need to include the version as part of the commit as PDE doesn't run those "fixes" in regular Tycho build, does it?

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java (hunk 1 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java]: Fixed in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/4685cbd46790097615112e789598bf0315e1d3a8.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java (hunk 2 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java]: Thanks, I updated it in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/4685cbd46790097615112e789598bf0315e1d3a8 to manually set the package export version in the manifest and removed the package-info.java for now.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java (hunk 3 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java]: Fixed in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/4685cbd46790097615112e789598bf0315e1d3a8.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java (hunk 2 above)
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/package-info.java]: > Good to know that modification can be automated, but still the MANIFEST.MF does need to include the version as part of the commit as PDE doesn't run those "fixes" in regular Tycho build, does it?

Yes, that currently only a very rough first step to empower further enhancements.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Implementations of this OSGi service interface allow to create new projects based on Maven archetypes.
+ * 
+ * @since 2.1.0 (package version 1.0.0)
+ */
+@ProviderType
+public interface ArchetypeGenerator {
+
+  /**
+   * Creates project structure using the given archetype and then imports created project(s)
```
Comment (kwin) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:38]: This javadoc (although just copied from ArchetypeGeneratorImpl) seems to require an update, as projects are no longer created by this method.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  default Collection<MavenProjectInfo> createArchetypeProjects(IPath location, IArchetype archetype, String groupId,
+      String artifactId, String version, String javaPackage, Properties properties, IProgressMonitor monitor)
+      throws CoreException
+  {
+    return createArchetypeProjects(location, archetype, groupId, artifactId, version, javaPackage, properties, false,
+        monitor);
+  }
+
+  Collection<MavenProjectInfo> createArchetypeProjects(IPath location, IArchetype archetype, String groupId,
+      String artifactId, String version, String javaPackage, Properties properties, boolean interactive,
```
Comment (HannesWell) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:51]: As proposed in #949 for the existing implementation we should consider to use a Map<String, String> instead of Properties. With static Map factories those are much more convenient to handle than Properties if there is a fixed set of key-values.

**Comment by HannesWell** [comment]
@kwin if you rebase on top of the current master, some (maybe all) of the CI failures should vanish.

**Comment by kwin** [comment]
Before I rebase and adjust the tests I would like to know if you are fine with the proposed API exposure.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java (hunk 5 above)
Comment (mickaelistria) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:51]: +1, `Map<String, String>` is more modern.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java [hunk 6]
```diff
@@ -11,16 +11,21 @@
...
 import org.eclipse.m2e.core.project.IArchetype;
 
 
-public class MavenArchetype implements IArchetype {
+/**
+ * Implementation of {@link IArchetype} based on {@link Archetype}
+ * 
+ * @since 2.1.0 (package version 1.0.0)
+ */
+public final class MavenArchetype implements IArchetype {
```
Comment (mickaelistria) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java:26]: Why is this required as API? I don't see it used in the signatures of ArchetypeManager/ArchetypeGenerator

**Review by mickaelistria — COMMENTED** [review]
This looks fresher and a well focused API! I just have a couple of questions/comments inline.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java (hunk 6 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java:26]: IArchetype is used as method parameter, so someone would need to come up with an implementation.

Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java:26]: See my comment, beside that a custom implementation is really trivial...

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java [hunk 7]
```diff
@@ -74,8 +77,8 @@
  *
  * @author Eugene Kuleshov
  */
-@Component(service = ArchetypePlugin.class)
-public class ArchetypePlugin {
+@Component(service = {ArchetypeManager.class, ArchetypePlugin.class})
```
[the remark this answers is not in the document]
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java:80]: Not possible, as the ArchetypePlugin is the internal API which is used. If you would only register as ArchetypePlugin you would need to cast in M2EUIPluginActivator. How do you want to get hold of this without casting?

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  List<RequiredProperty> getRequiredProperties(IArchetype archetype, IProgressMonitor monitor) throws CoreException;
+
+  /**
+   * Updates the local catalog to either add or refresh the given archetype.
+   * 
+   * @param archetype the archetype to add or refresh in the local catalog
+   * @throws CoreException
+   */
+  void updateLocalCatalog(Archetype archetype) throws CoreException;
```
[the remark this answers is not in the document]
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:57]: IArchetype does not give you the possibility to reference the artifact among other things. Why do you consider org.apache.maven.archetype.catalog.Archetype not suitable here?

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java (hunk 5 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java]: fixed by rebasing.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java (hunk 6 above)
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java]: Done in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/c39ab6b3ee2ade42b2a7da710f82006791c1c38b

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java [hunk 9]
```diff
@@ -48,11 +48,12 @@
 import org.eclipse.m2e.core.project.IProjectConfigurationManager;
 import org.eclipse.m2e.core.project.LocalProjectScanner;
 import org.eclipse.m2e.core.project.MavenProjectInfo;
+import org.eclipse.m2e.core.ui.archetype.ArchetypeGenerator;
 
 
-@Component(service = {ArchetypeGenerator.class})
+@Component(service = {ArchetypeGeneratorImpl.class})
```
[the remark this answers is not in the document]
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGeneratorImpl.java]: Fixed in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/c39ab6b3ee2ade42b2a7da710f82006791c1c38b.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:57]: Then this is probably not suitable for an API method here, why does user code needs this? The inital proposal in the issue was that it should be possible to generate a project given an Archetype (that is GAV) updating catalogs seems not something user code work with.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java:80]: > DS allows to replace one service implementation with another one!
This has nothing to do with DS, anyways you would be most likely wrecked then anyways. If you want be extra you can assert the type with instanceof and throw an exception otherwise...

> Also remember that you previously registered only to the internal API ArchetypePlugin via DS, so that pattern is not something I introduced.

Many thing in m2e need improvements and some are just for migrating thing so we should not take this as good examples when refactoring/rework things.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java (hunk 6 above)
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/MavenArchetype.java]: > Done in [c39ab6b](https://github.com/eclipse-m2e/m2e-core/commit/c39ab6b3ee2ade42b2a7da710f82006791c1c38b)

This looks good to me!

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (kwin) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java]: No longer necessary, removed from API in https://github.com/eclipse-m2e/m2e-core/pull/922/commits/50971b871cd650381d5744a9d114de1f3817aa87

[the remark this answers is not in the document]
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java]: Great :+1:

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypePlugin.java:80]: In Softwaredevelopment "later" often means "never" :-)

I would prefer to have a cast on the cases where it is required (that is we use no API method) as this is already done elsewhere in m2e and shows where we have an API gap...

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+import org.eclipse.m2e.core.project.IArchetype;
+
+
+/**
+ * Provides operations to manage the catalog.
+ *
+ * @since 2.1.0 (package version 1.0.0)
+ */
+@ProviderType
+public interface ArchetypeManager {
```
Comment (laeubi) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:37]: Why is ArchetypeManager public? Should ArchetypeGenerator not be enough?

**Comment by laeubi** [comment]
Looks good so far, just wondering if we really need to make `ArchetypeManager` public?

**Comment by laeubi** [comment]
> ArchetypeManager is used e.g. 

But all these should be m2e internal, so can we not make `ArchetypeManager` internal as well?

**Comment by laeubi** [comment]
> I use it outside of m2e, so I prefer using it with an API. I don't want to depend on internal API for my use case which was the original driver for this PR.

Can you explain how you use this? Do you provide own UI wizard? The main problem with this (and why it was decided to make this "non API") is that the Archetype is a separate plugin, so exposing its classes directly has two problems:

1. We need to make people depend on a specific ArcheType version (that is the one embedded in m2e) and we only support it partially, e.g. some classes will throw CNF because we only embedd the bar minimum of stuff.
2. If we update the ArcheType version and that API changes, this is actually an API break for m2e even though nothing was changed here.

This has lead to the situation that previously m2e a very long time relied on outdated Archetype 2.x. and it was quite hard to upgrade to 3.x. The result is/was the `IArcheType` facade interface that exposed the bare minimum required.

Can you check if for your use-case something similar would be possible to have (read only) facade types with a reduced signature for `RequiredProperty` and `ArchetypeCatalog`?

**Comment by laeubi** [comment]
About `ArchetypeParametersWizardPage` do you think you probably would reuse the m2e page? It also supports parameter validation through regexp now, otherwhise I think we can have an `IRequiredProperty` facade that wraps the maven `RequiredProperty`with just the getters required here.

**Comment by TIBCOeddie** [comment]
Our product uses the ArchetypeGenerator and friends. Is there still time to look at something and provide comments?

**Comment by laeubi** [comment]
I think best would be too finish this PR and then if there are still things missing start with a new PR.

**Comment by HannesWell** [comment]
> I think best would be too finish this PR and then if there are still things missing start with a new PR.

If it directly about what is changed in this PR I think comments should be made now. Of course it can be started again afterwards, but if possible why not consider them now but later?
If it is more an extension I agree that doing it step by step is better. :)

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ * Implementations of this OSGi service interface allow to create new Maven projects based on Maven archetypes.
+ * 
+ * @since 2.1.0 (package version 1.0.0)
+ */
+@ProviderType
+public interface ArchetypeGenerator {
+
+  /**
+   * Creates a project structure using the given archetype in non-interactive mode.
+   *
```
Comment (TIBCOeddie) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:39]: please have javadoc for params, or you'll get all the javadoc warnings especially since j17+ have turned it up, and for sure it's always a pity to not have these! esp handy is what can be null, what is optional, etc.

Comment (TIBCOeddie) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeGenerator.java:39]: I remember our integration had to dive into source to understand what "interactive mode" was

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+  /**
+   * Gets the required properties of an {@link IArchetype}.
+   *
+   * @param archetype the archetype possibly declaring required properties
+   * @param remoteArchetypeRepository the remote archetype repository, can be null
```
Comment (TIBCOeddie) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:43]: param is not in sync with actual

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+   * @param monitor the progress monitor, can be null
+   * @return the required properties of the archetypes, null if none is found
```
Comment (TIBCOeddie) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:45]: when your signature has a list, isn't it eclipse code guidelines to not return null and return an empty list instead?

Comment (mickaelistria) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:45]: +1, it's generally a good practice to not return null for Arrays or Collections but empty ones; not only for Eclipse but for Java in general.

Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:45]: Better return an `Optional<...>` instead then. There is a difference between an Archetype not found and having no required properties... one might even throw a CoreException if not found here ...

Comment (mickaelistria) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:45]: > Better return an Optional<...> instead then.

I don't think so.

>  There is a difference between an Archetype not found and having no required properties..

Maybe, but that's not the goal of this method to detect that an Archetype is not found.  But your comment here emphasize that documentation is ambiguous and that there are different ways to interpret `none`.

> one might even throw a CoreException if not found here ...

CoreException is just an abstract exception in Eclipse world; creating a custom exception type would be better for an API.

Comment (laeubi) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/archetype/ArchetypeManager.java:45]: > > Better return an Optional<...> instead then.
> 
> I don't think so.

Why? Either is is optional (because it is fine / could happen) but then it is not valid to just return an empty list if nothing is found as a user of that method then falsely assumes everything is fine and one could potentially call this archetype (what will just fail because it is not found)...

> Maybe, but that's not the goal of this method to detect that an Archetype is not found. But your comment here emphasize that documentation is ambiguous and that there are different ways to interpret none.

Obviously there is a way to detect this, so either it is an error (= exception must be thrown) or a user has to be able to distinguish the cases.

**Comment by mickaelistria** [comment]
> Why? Either is is optional (because it is fine / could happen) but then it is not valid to just return an empty list if nothing is found as a user of that method then falsely assumes everything is fine and one could potentially call this archetype (what will just fail because it is not found)...

There is a lot of literature against null collections, the most "reference" one being the "Effective Java" book, Your comment is based IMO on a dogmatic vision of API development; but experience of the whole Java (and other languages as well) community has concluded that returning null collections is more an anti-pattern than a benefit.

> Obviously there is a way to detect this, so either it is an error (= exception must be thrown) or a user has to be able to distinguish the cases.

Or maybe we just don't care about distinguishing, do we? Or maybe since we have an `Archetype` object as parameter, then we can assume that this Archetype exists and is found already.

**Comment by laeubi** [comment]
> There is a lot of literature against null collections,

And how does it relates to returning an empty Optional in case nothing could be computed here?

> Or maybe we just don't care about distinguishing, do we? Or maybe since we have an `Archetype` object as parameter, then we can assume that this Archetype exists and is found already.

The parameter is just a simple container object and obviously the creator of that method has taken into account that the archetype is not present.

**Comment by mickaelistria** [comment]
Given the comments, it seems like the overall design of those APIs is then questionable as they come with very poor guarantee about validity and lifecycle of instances.

So the current IArchetype is not an IArchetype then, it's just a GAV which is more used as an ArchetypeRequest. If I have an IArchetype, in a good API, I expect the archetype exists. The issue is then not what to do with a non existing IArchetype, it's that the API should ensure an IArchetype maps to something existing.
In the current state, I'd suggest removing the IArchetype.of method which is just a GAV, and even deprecating the IArchetype interface; or to make the IArchetype interface host the method to query parameters and other stuff and let only the ArchetypeManager create the IArchetype instance if this one does exists.

**Comment by HannesWell** [comment]
What's the status here?

**Comment by laeubi** [comment]
@kwin please rebase and resolve conflicts.

**Comment by laeubi** [comment]
Closing this for now as no progress and no response, but feel fre to just reopen or provide a new PR on the top of the master branch.
